File: .leetcode/647.回文子串.cpp
/*
 * @lc app=leetcode.cn id=647 lang=cpp
 *
 * [647] 回文子串
 */

// @lc code=start
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int countSubstrings(string s) {
        int length = s.length();
        int ans = 0;
        vector<vector<bool>>dp(length,vector<bool>(length,false));
        for(int i = length - 1 ; i >= 0;i--){
            for(int j = i;j<length;j++){
                if(s[i]==s[j]){
                    if(j - i<=1)
                        dp[i][j]=true;
                    else{
                        dp[i][j] = dp[i+1][j-1];
                    }
                    ans = dp[i][j]?ans+1:ans;
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: .leetcode/647.回文子串.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=647 lang=java
 *
 * [647] 回文子串
 */

// @lc code=start
class Solution {
    public int countSubstrings(String s) {
        int n = s.length();
        boolean[][] dp =new boolean[n + 1][n + 1];
        int ans = n;
        for(int i = 0;i <= n;i++){
            for(int j = i;j >= 0;j--){
                dp[i][j] = true;
            }
        }
        for(int i = 1;i <= n;i++){
            for(int j = 1;i + j <= n;j++){
                if(s.charAt(j - 1) == s.charAt(i + j - 1)){
                    dp[j][i + j] = dp[j + 1][i + j - 1];
                    if(dp[j][i + j])ans++;
                }else{
                    dp[j][i + j] = false;
                }
            }
        }
        return ans;
    }
}
// @lc code=end

